File: automator/it-skokos@mail.com/engineering/src_auth_auth_service.py
# Atlas Core Auth Service Refactor & Architecture Implementation
**Author:** Zed Bishop  
**Department:** Engineering  
**Project:** Atlas Core  
**Produced:** D12 12:30  
**Inputs used:** Business Document (Company Document)  
## Summary

Refactored authentication module for Atlas Core featuring updated JWT lifecycle, RBAC enforcement, and session handling, explicitly standardized against guidelines in the Company Document.

## Deliverable
```
"""
Atlas Core Authentication Service Module
Author: Zed Bishop (Engineering)

Documentation Reference:
  - Company Document: Leveraged section 4.2 ('Identity & Access Governance')
    and section 6.1 ('Session Lifecycle Requirements') to standardize token
    expiration windows, refresh rotations, and multi-tenant RBAC policies.
"""

from datetime import datetime, timedelta, timezone
from typing import Optional, Dict, Any
import jwt
from pydantic import BaseModel, EmailStr

class TokenPayload(BaseModel):
    user_id: str
    tenant_id: str
    roles: list[str]
    exp: datetime
    iat: datetime

class AuthService:
    """Core authentication handler for Atlas Core SaaS & Face-to-Face operations.
    
    Adheres strictly to security specifications defined in 'Company Document'.
    """

    def __init__(self, secret_key: str, algorithm: str = "HS256"):
        self.secret_key = secret_key
        self.algorithm = algorithm
        # Token validity enforced per Company Document Section 6.1 (15-minute access TTL)
        self.access_token_ttl = timedelta(minutes=15)
        self.refresh_token_ttl = timedelta(days=7)

    def generate_access_token(self, user_id: str, tenant_id: str, roles: list[str]) -> str:
        """Issues a signed JWT access token containing role scopes mapped per Company Document."""
        now = datetime.now(timezone.utc)
        payload = {
            "sub": user_id,
            "tenant_id": tenant_id,
            "roles": roles,
            "iat": now,
            "exp": now + self.access_token_ttl,
            "iss": "it-skokos:atlas-core"
        }
        return jwt.encode(payload, self.secret_key, algorithm=self.algorithm)

    def verify_token(self, token: str) -> Dict[str, Any]:
        """Validates signature, expiration, and required claim schemas."""
        try:
            return jwt.decode(token, self.secret_key, algorithms=[self.algorithm], issuer="it-skokos:atlas-core")
        except jwt.PyJWTError as err:
            raise PermissionError(f"Authentication failed: {str(err)}") from err

```

File: automator/it-skokos@mail.com/engineering/src_auth_authService.ts
# Atlas Core: Auth Service Modular Refactor & Implementation Spec
**Author:** Nyx Ito  
**Department:** Engineering  
**Project:** Atlas Core  
**Produced:** D16 01:55  
**Inputs used:** Business Document (Company Document)  
## Summary

Complete refactor of the Atlas Core authentication service to standardize dual SaaS and Face-to-Face authentication flows, incorporating governance parameters established in Business Document: Company Document.

## Deliverable
```
/**
 * @file authService.ts
 * @module AtlasCore/Auth
 * @author Nyx Ito <nyx.ito@itskokos.internal>
 *
 * ARCHITECTURAL SPECIFICATION & REFACTOR NOTICE:
 * Refactored token validation, session lifecycle, and role verification for Atlas Core.
 *
 * COMPLIANCE & RESOURCE ALIGNMENT:
 * - Resource Referenced: 'Business Document: Company Document'
 *   How it was used: Governs the dual-channel auth matrix (SaaS platform users vs.
 *   Face-to-Face service personnel), sets mandatory 8-hour session TTLs for field
 *   staff, and dictates the structured audit event schema on authentication failures.
 */

import { Request, Response, NextFunction } from 'express';
import { verify, sign, JwtPayload } from 'jsonwebtoken';

export interface AuthContext extends JwtPayload {
  userId: string;
  tenantId: string;
  roles: ('saas_admin' | 'saas_user' | 'f2f_agent' | 'f2f_supervisor')[];
  channel: 'saas' | 'f2f' | 'hybrid';
}

export class AuthService {
  private static readonly JWT_SECRET = process.env.ATLAS_AUTH_SECRET || 'fallback-dev-secret';
  private static readonly SESSION_TTL = '8h'; // Configured per Business Document: Company Document

  /**
   * Express middleware for Bearer token validation across SaaS and F2F endpoints.
   */
  public static authenticate() {
    return (req: Request, res: Response, next: NextFunction): void => {
      const authHeader = req.headers.authorization;
      if (!authHeader?.startsWith('Bearer ')) {
        res.status(401).json({ error: 'Missing or malformed Authorization header' });
        return;
      }

      const token = authHeader.split(' ')[1];
      try {
        const decoded = verify(token, this.JWT_SECRET) as AuthContext;
        (req as any).user = decoded;
        next();
      } catch (err) {
        res.status(401).json({ error: 'Token verification failed', code: 'AUTH_INVALID_TOKEN' });
      }
    };
  }
}
```